### DISCUSSION TITLE
Better Font Support

### DISCUSSION THREAD

**Discussion by supthos** [microsoft/WindowsAppSDK#4607] [Q&A]
Hi!
I'm developing an app to produce custom typefaces with the Windows App SDK. You all promised rendering with DWriteCore, but at the moment, v. 1.5, this is impossible with custom fonts, loaded at runtime with a file picker, for example. 
The only workaround I have found to possibly work, but not implemented yet is using Win2D Canvas. But that would mean that having to load the font through that library, instead of through DWriteCore.
Also, even if I wanted to render with DWriteCore, there is no clear way to, for example, render a thumbnail to an image without having to go through a device context like a window. Or maybe I'm just not informed on how to do this properly.
Lastly, I need better documentation and examples. The example provided for rendering with DWriteCore is unusable at best and very obfuscated, and does not follow the project structure for building a Windows App SDK app in Visual Studio, but seems to follow the paradigms of Win32 C++ more common with console application logic. Let's not talk about character encoding.
Please, help me out.... I've been stuck trying to render custom fonts for months now. It shouldn't be this complicated!

**Comment by castorix** [comment]
> this is impossible with custom fonts, loaded at runtime with a file picker, for example.

You follow MSDN at :
https://learn.microsoft.com/en-us/windows/win32/directwrite/custom-font-sets-win10#creating-a-font-set-using-arbitrary-fonts-in-the-local-file-system

**Comment by castorix** [comment]
I made a quick test without _IDWriteFontSetBuilder_ to simplify, with some random fonts I downloaded from Google, by calling
_CreateFontFileReference_/_CreateFontFace_ with a Path Geometry to draw a gradient from glyph indices with a _SwapChainPanel_ and it seems to work correctly (Windows App SDK 1.5.24, in C# because I cannot compile C++/WinRT) 

Test with "Lovely Home" font :

![image](https://github.com/user-attachments/assets/8b54152a-932d-452e-aeb6-e37e4f14f197)

**Comment by castorix** [comment]
>  Also, even if I wanted to render with DWriteCore, there is no clear way to, for example, render a thumbnail to an image without having to go through a device context like a window. Or maybe I'm just not informed on how to do this properly. Lastly, I need better documentation and examples

I just uploaded a test sample in C#, Windows App SDK 1.5.24 : [WinUI3_SwapChainPanel_DWriteCore](https://github.com/castorix/WinUI3_SwapChainPanel_DWriteCore
)
To be improved : as I added in comments, I met a problem with quality vs DWrite for _IDWriteTextLayout_
(I let _DWriteCreateFactory_ in comments to switch DWrite/DWriteCore and also a _Render2_ function for tests with less code for rendering)
